File: dart/leet_code/205.dart
/// Dart implementation of the LeetCode problem 205: Isomorphic Strings.
///
/// This module contains a class `Solution` with a method `isIsomorphic`.
/// The `isIsomorphic` method takes in two parameters:
/// - `s`: A string representing the first input string.
/// - `t`: A string representing the second input string.
///
/// The method returns a boolean indicating whether the two input strings are isomorphic.
/// Two strings are isomorphic if the characters in `s` can be replaced to get `t`.
///
/// Error Handling:
/// - This method assumes that the input parameters are well-formed, i.e., `s` and `t` are strings.
/// - If the input parameters are not well-formed, the behavior of the method is undefined.

class Solution {
  /// Determines whether two strings are isomorphic.
  ///
  /// This method uses a map to keep track of the mapping from characters in `s` to characters in `t`,
  /// and a set to keep track of the characters in `t` that have already been mapped to.
  ///
  /// @param s The first input string.
  /// @param t The second input string.
  /// @return A boolean indicating whether the two input strings are isomorphic.
  bool isIsomorphic(String s, String t) {
    // If the lengths of the strings are not equal, they cannot be isomorphic.
    if (s.length != t.length) {
      return false;
    }

    // Initialize a map to keep track of the mapping from characters in `s` to characters in `t`.
    Map<String, String> map = {};

    // Initialize a set to keep track of the characters in `t` that have already been mapped to.
    Set<String> set = {};

    // Traverse the strings.
    for (int i = 0; i < s.length; i++) {
      String key = s[i];
      String value = t[i];

      // If the current character in `s` has already been mapped to a character in `t`,
      // check if it is mapped to the current character in `t`.
      if (map.containsKey(key)) {
        if (map[key] != value) {
          return false;
        }
      } else {
        // If the current character in `s` has not been mapped and the current character in `t` has already been mapped to,
        // the strings cannot be isomorphic.
        if (set.contains(value)) {
          return false;
        }

        // Map the current character in `s` to the current character in `t` and add the current character in `t` to the set.
        map[key] = value;
        set.add(value);
      }
    }

    // If all characters in `s` can be replaced to get `t`, the strings are isomorphic.
    return true;
  }
}

File: java/leet_code/205.java
import java.util.HashMap;
import java.util.HashSet;

/**
 * This class provides a solution for checking if two strings are isomorphic.
 * Two strings are isomorphic if the characters in the first string can be replaced to get the second string.
 * The problem is solved by using a hash map to keep track of the mapping from characters in the first string to characters in the second string,
 * and a hash set to keep track of the characters in the second string that have already been mapped to.
 * If a character in the first string has already been mapped to a different character in the second string, or if a character in the second string has already been mapped to,
 * the strings are not isomorphic.
 *
 * @author jerodg
 */
class Solution {

    /**
     * Checks if two strings are isomorphic.
     *
     * @param s The first input string. It is a string of characters.
     * @param t The second input string. It is a string of characters.
     *
     * @return A boolean indicating whether the two strings are isomorphic. It is true if the strings are isomorphic, and false otherwise.
     */
    public static boolean isIsomorphic(final String s, final String t) {
        // If the strings have different lengths, they are not isomorphic
        if (s.length() != t.length()) return false;

        // Initialize a hash map to keep track of the mapping from characters in the first string to characters in the second string
        final HashMap<Character, Character> map = new HashMap<>();
        // Initialize a hash set to keep track of the characters in the second string that have already been mapped to
        final HashSet<Character> set = new HashSet<>();

        // Iterate over the characters in the strings
        for (int i = 0; i < s.length(); i++) {
            final char c1 = s.charAt(i);
            final char c2 = t.charAt(i);

            // If the current character in the first string has already been mapped to a character in the second string
            if (map.containsKey(c1)) {
                // If the current character in the first string is not mapped to the current character in the second string, the strings are not isomorphic
                if (map.get(c1) != c2) return false;
            } else {
                // If the current character in the second string has already been mapped to, the strings are not isomorphic
                if (set.contains(c2)) return false;

                // Map the current character in the first string to the current character in the second string
                map.put(c1, c2);
                // Add the current character in the second string to the set of characters that have been mapped to
                set.add(c2);
            }
        }

        // If all characters in the first string can be replaced to get the second string, the strings are isomorphic
        return true;
    }
}
